test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/KeyAsSegmentTests/Tests/ClientEndToEndTests.cs

```diff
@@ -0,0 +1,286 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ClientEndToEndTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using Microsoft.AspNetCore.OData;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Client.E2E.TestCommon;
+using Microsoft.OData.Client.E2E.Tests.Common.Client.Default;
+using Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd;
+using Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Default;
+using Microsoft.OData.Client.E2E.Tests.Common.Server.EndToEnd;
+using Microsoft.OData.Client.E2E.Tests.KeyAsSegmentTests.Server;
+using Microsoft.OData.Edm;
+using Xunit;
+using Employee = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Employee;
+using Person = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Person;
+using SpecialEmployee = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.SpecialEmployee;
+using DiscontinuedProduct = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.DiscontinuedProduct;
+using ProductPhoto = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.ProductPhoto;
+using Customer = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Customer;
+using Order = Microsoft.OData.Client.E2E.Tests.Common.Clients.EndToEnd.Order;
+
+namespace Microsoft.OData.Client.E2E.Tests.KeyAsSegmentTests.Tests;
+
+public class ClientEndToEndTests : EndToEndTestBase<ClientEndToEndTests.TestsStartup>
+{
+    private readonly Uri _baseUri;
+    private readonly Container _context;
+    private readonly IEdmModel _model;
+
+    public class TestsStartup : TestStartupBase
+    {
+        public override void ConfigureServices(IServiceCollection services)
+        {
+            services.ConfigureControllers(typeof(KeyAsSegmentTestsController), typeof(MetadataController));
+
+            services.AddControllers().AddOData(opt =>
+                opt.EnableQueryFeatures().AddRouteComponents("odata", CommonEndToEndEdmModel.GetEdmModel()));
+        }
+    }
+
+    public ClientEndToEndTests(TestWebApplicationFactory<TestsStartup> fixture) : base(fixture)
+    {
+        if (Client.BaseAddress == null)
+        {
+            throw new ArgumentNullException(nameof(Client.BaseAddress), "Base address cannot be null");
+        }
+
+        _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+        _context = new Container(_baseUri)
+        {
+            HttpClientFactory = HttpClientFactory
+        };
+
+        _model = CommonEndToEndEdmModel.GetEdmModel();
+        ResetDefaultDataSource();
+    }
+
+    [Fact]
+    public void GetSingleEntity()

```

Comment:
Consider giving these methods a name that makes it clear what scenario/behaviour is being tested here. Or at least make the name of the class more specific.